### PR TITLE
Add support for delegation and return values with @yieldfrom

### PR DESCRIPTION
**Pull request by gerlero**
This is a basic implementation of delegation to a subgenerator inspired by [PEP 380](https://peps.python.org/pep-0380/). Delegation with a plain `@yieldfrom iter` expression (i.e., no assignment on the left) should work for any iterable, while `arg = @yieldfrom f` requires that `f` is actually a `@resumable` function, as I didn't find a standard mechanism for iterators to "return" values. Note that in the latter case, `arg` will be assigned the return value of `f`: this also means that it now makes sense for a `@resumable` function to also return a value, so I've removed the warning for this case.

(https://github.com/BenLauwens/ResumableFunctions.jl/pull/57)

EDIT: typo, rename `@yield_from` -> `@yieldfrom`

### CODE REVIEW

**Comment by gerlero** [comment]
Already took care of the merge conflicts caused by the package rename.

As an aside, I now think `@yieldfrom` _might_ be a more "Julian" name for the new macro. OTOH, `@yield_from` (the name I gave it in this PR) is more resembling of the Python `yield from` syntax that it attempts to replicate. Torn on this one.

**Comment by Krastanov** [comment]
Thanks! Not quite clear to me why codecov broke. I will try to fix it and then I will get back to the pull request

**Comment by hdavid16** [comment]
> As an aside, I now think `@yieldfrom` _might_ be a more "Julian" name for the new macro. OTOH, `@yield_from` (the name I gave it in this PR) is more resembling of the Python `yield from` syntax that it attempts to replicate. Torn on this one.

What about a syntax like `@yield from ...`? This would me another method for the current `@yield` macro. DiscreteEvents.jl uses `@enum` to create subtypes `after`, `every`, `at` for their `@event` macro: see  https://juliadynamics.github.io/DiscreteEvents.jl/dev/events/#Use-Cases

**Comment by Krastanov** [comment]
I am a bit reluctant to start creating new keywords (and generally do significant syntax magic and DSLs) at the start of the fork. I would like to wait a bit to (1) get more familiar with the code base and add a bit more static analysis and tests to it and (2) observe whether such keywords feel needed before adding them.

**Comment by gerlero** [comment]
> Thanks! Not quite clear to me why codecov broke. I will try to fix it and then I will get back to the pull request

Would a rebase on master help?

**Comment by Krastanov** [comment]
> Would a rebase on master help?

Not sure, but right now I think it is more of a misconfiguration issue on my end. I should be able to fix it tonight. I will also slightly reorganize the tests done here (I will push directly to this branch) as I am trying to set up JET and Aqua static analysis.

**Comment by hdavid16** [comment]
Yeah, I'd vote for `@yieldfrom`

File: src/macro.jl [hunk 1]
```diff
@@ -5,6 +5,13 @@ macro yield(expr=nothing)
   error("@yield macro outside a @resumable function!")
 end
 
+"""
+Macro if used in a `@resumable function` that delegates to `expr` otherwise throws an error.
+"""
+macro yield_from(expr=nothing)
+  error("@yield_from macro outside a @resumable function!")
+end
```
Comment (Krastanov) [inline src/macro.jl:13]: Could we add a `@test_throws` for this, to make sure future changes do not bork it up?

**Comment by Krastanov** [comment]
@hdavid16 @gerlero , in a separate pull request I added JET and Aqua static analysis and reorganized the tests in the SciML style. I fixed the merge conflicts that appeared here.

@hdavid16 , could you look through this and give your own review?

I would also like to suggest that we try to keep the pull request easy to backport to ResumableFunctions in case public development resumes for that package. Given that there is already a pull request for them, I am not too worried about that, but just wanted to make sure I have mentioned it.

If both of you prefer `@yieldfrom`, could you make that change?

Lastly, to both of you, if you could consider signing up as co-maintainers for the package (so that if I get in a car crash the package does not die yet again), please let me know -- I would love to add you to the github team.

**Comment by gerlero** [comment]
@Krastanov 

> I would also like to suggest that we try to keep the pull request easy to backport to ResumableFunctions in case public development resumes for that package. Given that there is already a pull request for them, I am not too worried about that, but just wanted to make sure I have mentioned it.

I agree. It should be easily doable as long as I don't rebase. https://github.com/BenLauwens/ResumableFunctions.jl/pull/57 works off the same branch so it's not mergeable there right now, but after this one is done I could try to get the branch back into some state that _can_ be merged there. Otherwise, someone could just cherry-pick the relevant changes

> If both of you prefer `@yieldfrom`, could you make that change?

I think `@yieldfrom` strikes the right balance of naming consistency and discoverability (esp. to those who are familiar with Python generators). I'll make the change.

> Lastly, to both of you, if you could consider signing up as co-maintainers for the package (so that if I get in a car crash the package does not die yet again), please let me know -- I would love to add you to the github team.

Drive safe and take care, but do feel free to add me.

File: src/macro.jl (hunk 1 above)
Comment (gerlero) [reply src/macro.jl:13]: Added a test for that (using `eval()`, which wraps the actual error in a `LoadError`; not sure if there's a better way)

**Comment by Krastanov** [comment]
Hi, @gerlero , @hdavid16 ,

It seems Ben is reviving ResumableFunctions, and presumably SimJulia as well. He also added me as a maintainer. I suggest the following:

1. I will go ahead and merge this here, in the fork, just to have it saved in one form. @gerlero , if you have the time at some point, please consider backporting it to the original repository.
2. Let's check with Ben about the process he prefers for maintenance of the original package. I would like to volunteer you as additional maintainers if that is ok. Could you let me know?
3. For the moment I will keep this fork in sync with any improvements in the original repo.

Let me know if you have any other suggestions or questions or worries!

**Comment by gerlero** [comment]
> 1. I will go ahead and merge this here, in the fork, just to have it saved in one form. @gerlero , if you have the time at some point, please consider backporting it to the original repository.

I'll backport this to ResumableFunctions.jl as soon as it is merged.

> 2. Let's check with Ben about the process he prefers for maintenance of the original package. I would like to volunteer you as additional maintainers if that is ok. Could you let me know?

Feel free to volunteer me there as well.

> 3. For the moment I will keep this fork in sync with any improvements in the original repo.
> 
> Let me know if you have any other suggestions or questions or worries!

The timing of the fork and the owner of original package showing up again just now looks to have been very unfortunate and will presumably leave us with two equivalent packages, at least in the short term. Still, thanks for your work in starting this fork, as development on ResumableFunctions.jl was effectively dead before yesterday. I'll keep using the Semicoroutines.jl fork for now (especially since ResumableFunctions still hasn't released the required patch), but I can also migrate back if that package is revived _and_ you decide the fork is no longer necessary. My point is, I don't know if duplicating the development/maintenance effort is worth it for this package, and as long as they provide what I need I'm fine with just having one or the other.

**Comment by Krastanov** [comment]
I agree, long term having two basicallly-the-same packages does not make sense. A couple of folks have already transitioned so I am committing to backport things by hand for the next year or two to make sure the original package is healthy. At that point I will myself send pull requests to dependents to move back to the original package if necessary (and if the original is healthy).

**Pull request merged**